source/s3_meta_request.c

```diff
@@ -842,20 +832,269 @@ static void s_s3_meta_request_send_request_finish_default(
+
+    struct aws_s3_request *request = user_data;
+    AWS_PRECONDITION(request);
+
+    struct aws_s3_meta_request *meta_request = request->meta_request;
+    AWS_PRECONDITION(meta_request);
+
+    if (error_code != AWS_ERROR_SUCCESS) {
+        AWS_LOGF_ERROR(
+            AWS_LS_S3_META_REQUEST,
+            "id=%p Meta request could not get retry token request %p due to error %d (%s)",
+            (void *)meta_request,
+            (void *)request,
+            error_code,
+            aws_error_str(error_code));
+
+        goto error_ref_clean_up;
+    }
+
+    AWS_ASSERT(token);
+
+    enum aws_retry_error_type error_type = AWS_RETRY_ERROR_TYPE_TRANSIENT;
+
+    if (error_code == AWS_ERROR_S3_NO_PART_BUFFER) {
+        error_type = AWS_RETRY_ERROR_TYPE_CLIENT_ERROR;
+    } else if (error_code == AWS_ERROR_S3_INTERNAL_ERROR) {
+        error_type = AWS_RETRY_ERROR_TYPE_SERVER_ERROR;
+    } else if (error_code == AWS_ERROR_S3_SLOW_DOWN) {
+        error_type = AWS_RETRY_ERROR_TYPE_THROTTLING;
+    } else {
+        error_type = AWS_RETRY_ERROR_TYPE_TRANSIENT;
+    }
+
+    request->retry_token = token;
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_S3_META_REQUEST,
+        "id=%p Meta request scheduling retry of request %p with token %p",
+        (void *)meta_request,
+        (void *)request,
+        (void *)token);
+
+    /* Ask the retry strategy to schedule a retry of the request. */
+    if (aws_retry_strategy_schedule_retry(request->retry_token, error_type, s_s3_meta_request_retry_ready, request)) {
+        error_code = aws_last_error();
+
+        AWS_LOGF_ERROR(
+            AWS_LS_S3_META_REQUEST,
+            "id=%p Meta request could not retry request %p with token %p due to error %d (%s)",
+            (void *)meta_request,
+            (void *)request,
+            (void *)token,
+            error_code,
+            aws_error_str(error_code));
+
+        goto error_ref_clean_up;
+    }
+
+    return;
+
+error_ref_clean_up:
+
+    aws_s3_request_release(request);
+    aws_s3_meta_request_release(meta_request);
+
+    aws_s3_meta_request_finish(meta_request, NULL, 0, aws_last_error());
+}
+
+static void s_s3_meta_request_acquire_retry_token(
+    struct aws_retry_strategy *retry_strategy,
+    struct aws_s3_request *request) {
+    AWS_PRECONDITION(retry_strategy);
+    AWS_PRECONDITION(request);
+
+    struct aws_s3_meta_request *meta_request = request->meta_request;
+    AWS_PRECONDITION(meta_request);
+
+    struct aws_http_message *message = meta_request->initial_request_message;
+    AWS_PRECONDITION(message);
+
+    struct aws_byte_cursor host_header_name = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("Host");
+
+    struct aws_http_headers *message_headers = aws_http_message_get_headers(message);
+    AWS_ASSERT(message_headers);
+
+    struct aws_byte_cursor host_header_value;
+
+    bool host_header_exists =
+        aws_http_headers_get(message_headers, host_header_name, &host_header_value) == AWS_ERROR_SUCCESS;
```

Comment:
maybe just if (blah) { log, goto error_ref_clean_up} like the below failure?

Reply:
We could--we should be guaranteed a host header in the passed in message though. :-/